Enable multipart/related on FileUpload

mufasa1976 opened the pull request:
SOAP-Requests with MTOM enabled are sending the Requests with Content-Type `multipart/related`.
Because the Parts would not have `content-disposition` Headers they would not be recognized by the FileUploads as Parts.

Because this Project is used by [Wiremock](https://github.com/wiremock/wiremock) it would fix the Issue https://github.com/wiremock/wiremock/issues/2176

mufasa1976 committed:
added ability to use content-type: multipart/related

mufasa1976 committed:
Merge branch 'apache:master' into master

mufasa1976 committed:
added entry to src/changes/changes.xml; removed unnecessary paragraph at javadoc

garydgregory reviewed (changes requested):
Hello @mufasa1976 
Please run 'mvn' locally before pushing and fix build issues.

mufasa1976 committed:
fixing checkstyle failures

mufasa1976 commented:
I've fixed the Checkstyle Bugs. But now I got a failure from Jacoco.
```
[WARNING] Rule violated for bundle commons-fileupload2-core: lines covered ratio is 0.41, but expected minimum is 0.42
```
The Problem is that I added a Test for the `FileItemInputIteratorImpl` and `AbstractFileUpload` but these will not be execute by the module `commons-fileupload2-core`. Indeed the changed Lines will be covered by Tests of all other Modules because these will extend `AbstractFileUploadTest` located within the Module `commons-fileupload2-core`.

But this "Warning" will cause that the Code Coverage of the Module `commons-fileupload2-core` sinks beyound 0.42 (indeed it is now 0.41). Should I set the Value of the Code-Coverage to 0.41 within `commons-fileupload2-core/pom.xml`?

garydgregory commented on commons-fileupload2-core/src/main/java/org/apache/commons/fileupload2/core/AbstractRequestContext.java:
```diff
@@ -20,8 +20,14 @@
 
 public abstract class AbstractRequestContext<T> implements RequestContext {
+    /**
+     * the Content-Type Pattern for multipart/related Requests.
```
Sentences start with a capital letter.

garydgregory commented on commons-fileupload2-core/src/main/java/org/apache/commons/fileupload2/core/AbstractRequestContext.java:
```diff
@@ -79,4 +85,7 @@ public String toString() {
         return String.format("%s [ContentLength=%s, ContentType=%s]", getClass().getSimpleName(), getContentLength(), getContentType());
     }
 
+    protected boolean isMultipartRelated(final String contentType) {
```
New protected and public items should have a Javadoc and `since` tag.

garydgregory commented on commons-fileupload2-core/src/main/java/org/apache/commons/fileupload2/core/FileItemInputIteratorImpl.java:
```diff
@@ -96,6 +95,11 @@ class FileItemInputIteratorImpl implements FileItemInputIterator {
 
+    /**
+     * is the Request of type <code>multipart/related</code>.
+     */
```
Sentences start with a capital letter.

garydgregory commented on commons-fileupload2-core/src/main/java/org/apache/commons/fileupload2/core/RequestContext.java:
```diff
@@ -70,4 +70,10 @@ default Charset getCharset() throws UnsupportedCharsetException {
+     *
+     * @return the Request is of type <code>multipart/related</code>
+     */
+    boolean isMultipartRelated();
```
Add `since` Javadoc tag.

garydgregory commented on commons-fileupload2-jakarta-servlet5/src/main/java/org/apache/commons/fileupload2/jakarta/servlet5/JakartaServletRequestContext.java:
```diff
@@ -68,4 +68,13 @@ public InputStream getInputStream() throws IOException {
+     *
+     * @return the Request is of type <code>multipart/related</code>
+     */
+    @Override
```
Add `since` Javadoc tag.

garydgregory commented on commons-fileupload2-jakarta-servlet6/src/main/java/org/apache/commons/fileupload2/jakarta/servlet6/JakartaServletRequestContext.java:
```diff
@@ -68,4 +68,13 @@ public InputStream getInputStream() throws IOException {
+     * Is the Request of type <code>multipart/related</code>?
+     *
+     * @return the Request is of type <code>multipart/related</code>
+     */
```
Add `since` Javadoc tag.

garydgregory commented on commons-fileupload2-javax/src/main/java/org/apache/commons/fileupload2/javax/JavaxServletRequestContext.java:
```diff
@@ -68,4 +68,13 @@ public InputStream getInputStream() throws IOException {
+     *
+     * @return the Request is of type <code>multipart/related</code>
+     */
+    @Override
```
Add `since` Javadoc tag.

garydgregory commented on commons-fileupload2-portlet/src/main/java/org/apache/commons/fileupload2/portlet/JavaxPortletRequestContext.java:
```diff
@@ -68,4 +68,13 @@ public InputStream getInputStream() throws IOException {
+     * Is the Request of type <code>multipart/related</code>?
+     *
+     * @return the Request is of type <code>multipart/related</code>
+     */
```
Add `since` Javadoc tag.

garydgregory reviewed (changes requested):
@mufasa1976 
Thank you for your PR. 
Please see my comments.

garydgregory commented:
> I've fixed the Checkstyle Bugs. But now I got a failure from Jacoco.
> 
> ```
> [WARNING] Rule violated for bundle commons-fileupload2-core: lines covered ratio is 0.41, but expected minimum is 0.42
> ```
> 
> The Problem is that I added a Test for the `FileItemInputIteratorImpl` and `AbstractFileUpload` but these will not be execute by the module `commons-fileupload2-core`. Indeed the changed Lines will be covered by Tests of all other Modules because these will extend `AbstractFileUploadTest` located within the Module `commons-fileupload2-core`.
> 
> But this "Warning" will cause that the Code Coverage of the Module `commons-fileupload2-core` sinks beyound 0.42 (indeed it is now 0.41). Should I set the Value of the Code-Coverage to 0.41 within `commons-fileupload2-core/pom.xml`?



> I've fixed the Checkstyle Bugs. But now I got a failure from Jacoco.
> 
> ```
> [WARNING] Rule violated for bundle commons-fileupload2-core: lines covered ratio is 0.41, but expected minimum is 0.42
> ```
> 
> The Problem is that I added a Test for the `FileItemInputIteratorImpl` and `AbstractFileUpload` but these will not be execute by the module `commons-fileupload2-core`. Indeed the changed Lines will be covered by Tests of all other Modules because these will extend `AbstractFileUploadTest` located within the Module `commons-fileupload2-core`.
> 
> But this "Warning" will cause that the Code Coverage of the Module `commons-fileupload2-core` sinks beyound 0.42 (indeed it is now 0.41). Should I set the Value of the Code-Coverage to 0.41 within `commons-fileupload2-core/pom.xml`?



> Author
> ###

The point of the Jacoco check is to avoid making things worse and, conversely, to improve the overall quality of the component. So I encourage you to improve the state of the code.

Run: `mvn clean site` (or `mvn clean site -P jacoco`) and look at the Jacoco report in `target/site` where you'll need a nice HTML report with code highlights of what's tested and what's not. This will let you make sure that the code you touch in a PR is indeed touched by tests, new or existing. It's not perfect, but we want to quality overall, to grow, not shrink ;-)

mufasa1976 committed:
fixed findings of review

mufasa1976 commented:
> > I've fixed the Checkstyle Bugs. But now I got a failure from Jacoco.
> > ```
> > [WARNING] Rule violated for bundle commons-fileupload2-core: lines covered ratio is 0.41, but expected minimum is 0.42
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The Problem is that I added a Test for the `FileItemInputIteratorImpl` and `AbstractFileUpload` but these will not be execute by the module `commons-fileupload2-core`. Indeed the changed Lines will be covered by Tests of all other Modules because these will extend `AbstractFileUploadTest` located within the Module `commons-fileupload2-core`.
> > But this "Warning" will cause that the Code Coverage of the Module `commons-fileupload2-core` sinks beyound 0.42 (indeed it is now 0.41). Should I set the Value of the Code-Coverage to 0.41 within `commons-fileupload2-core/pom.xml`?
> 
> > Author
> 
> The point of the Jacoco check is to avoid making things worse and, conversely, to improve the overall quality of the component. So I encourage you to improve the state of the code.
> 
> Run: `mvn clean site` (or `mvn clean site -P jacoco`) and look at the Jacoco report in `target/site` where you'll see a nice HTML report with code highlights of what's tested and what's not. This will let you make sure that the code you touch in a PR is indeed touched by tests, new or existing. It's not perfect, but we want to quality overall, to grow, not shrink ;-)

I've ran `mvn clean site` (and also `mvn clean site -P jacoco`) but unfortunatly the Jacoco-Report of `commons-fileupload2-core` shows that **NO** Lines of `AbstractFileUpload` or `FileItemInputIteratorImpl` will be covered.

It seems that Jacoco doesn't take care of Project-internal Dependencies because `common-fileupload2-core` has been added as a compile Dependency to `common-fileupload2-servlet5`, `common-fileupload2-servlet6`, `common-fileupload2-javax` and `common-fileupload2-portlet`.

I've made a Code Coverage Run with IntelliJ on the Modules `common-fileupload2-servlet5`, `common-fileupload2-servlet6`, `common-fileupload2-javax` and `common-fileupload2-portlet` and I have seen that my Code will be covered by the Test I've written.

So unless there is a Chance to tell Jacoco to take some Dependencies in Content the only solution I see is to reduce the Line Complexity on `common-fileupload2-core` because some of these Classes (i.E. `AbstractFileUpload` or `FileItemInputIteratorImpl`) can't be covered by Tests on `common-fileupload2-core`

garydgregory commented:
Hello @mufasa1976 
Note that it's the lines covered ratio that is at issue, not the complexity.

You added runtime code like: 
`org.apache.commons.fileupload2.core.AbstractRequestContext.isMultipartRelated(String)` and its regular expression constant.
Since there is no concrete subclass in this module. create a subclass in the tests themselves and make sure the method does the right thing with the regular expression added in the PR.
And so on. This should allow the core module to not drop coverage. It should not rely on other modules to test its new code and regular expression anyway.

mufasa1976 committed:
added MockRequestContextTest

mufasa1976 committed:
added License Header

mufasa1976 commented:
> Hello @mufasa1976 Note that it's the lines covered ratio that is at issue, not the complexity.
> 
> You added runtime code like: `org.apache.commons.fileupload2.core.AbstractRequestContext.isMultipartRelated(String)` and its regular expression constant. Since there is no concrete subclass in this module. create a subclass in the tests themselves and make sure the method does the right thing with the regular expression added in the PR. And so on. This should allow the core module to not drop coverage. It should not rely on other modules to test its new code and regular expression anyway.

You're absolutly right. I've added a new Test `MockRequestContextTest` where I added additional Tests for `AbstractRequestContext`. And there I got a Coverage-Ratio of 0.42 or above.

mufasa1976 commented:
Thank you very much for your patience
